File: src/lib/test/MlKemInputCheckTests.cpp
/*****************************************************************************
 MlKemInputCheckTests.cpp

 Gap-closure plan 2026-09-25, finding E3 (C++ engine).

 Vectors: tests/acvp/mlkem_encapdecap_val_test.json — a byte-copied subset of
 NIST ACVP-Server@975de31e ML-KEM-encapDecap-FIPS203 (provenance in the
 file; copied from pqctoday-hub src/data/acvp/ @ 7656391c7).

 * Key checks (FIPS 203 §7.2 encapsulation-key modulus check, §7.3
   decapsulation-key hash check): C_CreateObject of every key-check case —
   testPassed=true -> CKR_OK, testPassed=false -> CKR_ATTRIBUTE_VALUE_INVALID
   (PKCS#11 v3.2 §4.1.1 rule 2; §6.68.2/§6.68.3 define CKA_VALUE as the FIPS
   203 ek/dk). Before the fix C_CreateObject stored all of them.
 * Decapsulation: every case (valid and modified ciphertext — implicit
   rejection) decapsulates byte-exact to the upstream k.
 * Boundaries: ek/dk one byte short or long -> CKR_ATTRIBUTE_VALUE_INVALID
   (FIPS 203 §7.2/§7.3 type check); a ciphertext of another parameter set's
   length (or off by one) -> CKR_WRAPPED_KEY_LEN_RANGE (PKCS#11 v3.2 §5.1.6
   "invalid solely on the basis of its length", listed for C_DecapsulateKey
   in §5.18.9), with no key object created. Before the fix a 768- or
   1568-byte ciphertext against an ML-KEM-768 key returned
   CKR_WRAPPED_KEY_INVALID and a 1-byte-short dk was accepted.
 *****************************************************************************/

#include <config.h>
#include "MlKemInputCheckTests.h"
#include "AcvpKatUtil.h"
#include <cstring>

CPPUNIT_TEST_SUITE_REGISTRATION(MlKemInputCheckTests);

namespace {

CK_ULONG mlkemParamSet(const std::string& name)
{
	if (name == "ML-KEM-512")  return CKP_ML_KEM_512;
	if (name == "ML-KEM-768")  return CKP_ML_KEM_768;
	if (name == "ML-KEM-1024") return CKP_ML_KEM_1024;
	return (CK_ULONG)-1;
}

const nlohmann::json* findCase(const nlohmann::json& doc, const char* function,
                               const char* paramSet, bool wantPassed)
{
	for (const auto& g : doc["testGroups"])
	{
		if (g["function"] != function || g["parameterSet"] != paramSet) continue;
		for (const auto& t : g["tests"])
			if (!t.contains("testPassed") || t["testPassed"].get<bool>() == wantPassed)
				return &t;
	}
	return NULL;
}

} // namespace

void MlKemInputCheckTests::openSession(CK_SESSION_HANDLE& hSession)
{
	CRYPTOKI_F_PTR( C_Finalize(NULL_PTR) );
	CK_RV rv = CRYPTOKI_F_PTR( C_Initialize(NULL_PTR) );
	CPPUNIT_ASSERT(rv == CKR_OK);
	rv = CRYPTOKI_F_PTR( C_OpenSession(m_initializedTokenSlotID, CKF_SERIAL_SESSION | CKF_RW_SESSION, NULL_PTR, NULL_PTR, &hSession) );
	CPPUNIT_ASSERT(rv == CKR_OK);
	rv = CRYPTOKI_F_PTR( C_Login(hSession, CKU_USER, m_userPin1, m_userPin1Length) );
	CPPUNIT_ASSERT(rv == CKR_OK);
}

CK_RV MlKemInputCheckTests::importKey(CK_SESSION_HANDLE hSession, bool isPrivate, CK_ULONG paramSet,
                                      const std::vector<unsigned char>& value, CK_OBJECT_HANDLE& hKey)
{
	CK_OBJECT_CLASS cls = isPrivate ? CKO_PRIVATE_KEY : CKO_PUBLIC_KEY;
	CK_KEY_TYPE kt = CKK_ML_KEM;
	CK_BBOOL bFalse = CK_FALSE, bTrue = CK_TRUE;
	std::vector<unsigned char> v(value);
	std::vector<CK_ATTRIBUTE> tmpl = {
		{ CKA_CLASS, &cls, sizeof(cls) },
		{ CKA_KEY_TYPE, &kt, sizeof(kt) },
		{ CKA_TOKEN, &bFalse, sizeof(bFalse) },
		{ (CK_ATTRIBUTE_TYPE)(isPrivate ? CKA_DECAPSULATE : CKA_ENCAPSULATE), &bTrue, sizeof(bTrue) },
		{ CKA_PARAMETER_SET, &paramSet, sizeof(paramSet) },
		{ CKA_VALUE, v.empty() ? NULL_PTR : v.data(), (CK_ULONG)v.size() },
	};
	if (isPrivate)
	{
		tmpl.push_back({ CKA_SENSITIVE, &bFalse, sizeof(bFalse) });
		tmpl.push_back({ CKA_EXTRACTABLE, &bTrue, sizeof(bTrue) });
	}
	hKey = CK_INVALID_HANDLE;
	return CRYPTOKI_F_PTR( C_CreateObject(hSession, tmpl.data(), (CK_ULONG)tmpl.size(), &hKey) );
}

CK_RV MlKemInputCheckTests::decapsulate(CK_SESSION_HANDLE hSession, CK_OBJECT_HANDLE hPrk,
                                        const std::vector<unsigned char>& ct, CK_OBJECT_HANDLE& hSecret)
{
	CK_MECHANISM mech = { CKM_ML_KEM, NULL_PTR, 0 };
	CK_OBJECT_CLASS secretClass = CKO_SECRET_KEY;
	CK_BBOOL bTrue = CK_TRUE, bFalse = CK_FALSE;
	CK_ATTRIBUTE tmpl[] = {
		{ CKA_CLASS, &secretClass, sizeof(secretClass) },
		{ CKA_TOKEN, &bFalse, sizeof(bFalse) },
		{ CKA_EXTRACTABLE, &bTrue, sizeof(bTrue) },
		{ CKA_SENSITIVE, &bFalse, sizeof(bFalse) },
	};
	std::vector<unsigned char> c(ct);
	hSecret = CK_INVALID_HANDLE;
	return CRYPTOKI_F_PTR( C_DecapsulateKey(hSession, &mech, hPrk, tmpl, sizeof(tmpl) / sizeof(CK_ATTRIBUTE),
	                                        c.data(), (CK_ULONG)c.size(), &hSecret) );
}

void MlKemInputCheckTests::testNistKeyChecks()
{
	CK_SESSION_HANDLE hSession;
	openSession(hSession);
	nlohmann::json doc = acvpkat::load("mlkem_encapdecap_val_test.json");
	acvpkat::Failures f;
	size_t negatives = 0;
	for (const auto& g : doc["testGroups"])
	{
		const std::string fn = g["function"].get<std::string>();
		if (fn != "encapsulationKeyCheck" && fn != "decapsulationKeyCheck") continue;
		const bool isPrivate = (fn == "decapsulationKeyCheck");
		const std::string ps = g["parameterSet"].get<std::string>();
		for (const auto& t : g["tests"])
		{
			const bool passed = t["testPassed"].get<bool>();
			if (!passed) negatives++;
			const std::string id = "tc" + std::to_string(t["tcId"].get<int>()) + " " + ps + " " + fn +
				" (" + t["reason"].get<std::string>() + ")";
			CK_OBJECT_HANDLE h;
			CK_RV rv = importKey(hSession, isPrivate, mlkemParamSet(ps),
			                     acvpkat::hexField(t, isPrivate ? "dk" : "ek"), h);
			const CK_RV expected = passed ? CKR_OK : CKR_ATTRIBUTE_VALUE_INVALID;
			f.check(rv == expected, id + ": C_CreateObject expected " + acvpkat::rvHex(expected) +
			        ", got " + acvpkat::rvHex(rv));
			if (!passed)
				f.check(h == CK_INVALID_HANDLE, id + ": a refused key must not yield a handle");
			if (rv == CKR_OK && passed && !isPrivate)
			{
				// A key that passed the check stays usable.
				CK_MECHANISM mech = { CKM_ML_KEM, NULL_PTR, 0 };
				CK_OBJECT_CLASS secretClass = CKO_SECRET_KEY;
				CK_ATTRIBUTE st[] = { { CKA_CLASS, &secretClass, sizeof(secretClass) } };
				CK_ULONG ctLen = 0;
				CK_OBJECT_HANDLE hs = CK_INVALID_HANDLE;
				CK_RV erv = CRYPTOKI_F_PTR( C_EncapsulateKey(hSession, &mech, h, st, 1, NULL_PTR, &ctLen, &hs) );
				f.check(erv == CKR_OK && ctLen > 0, id + ": C_EncapsulateKey size query rv=" + acvpkat::rvHex(erv));
			}
			if (h != CK_INVALID_HANDLE) CRYPTOKI_F_PTR( C_DestroyObject(hSession, h) );
		}
	}
	f.assertNone("mlkem_encapdecap_val_test.json key checks");
	CPPUNIT_ASSERT_EQUAL((size_t)6, negatives);
}

void MlKemInputCheckTests::testNistDecapsulation()
{
	CK_SESSION_HANDLE hSession;
	openSession(hSession);
	nlohmann::json doc = acvpkat::load("mlkem_encapdecap_val_test.json");
	acvpkat::Failures f;
	for (const auto& g : doc["testGroups"])
	{
		if (g["function"] != "decapsulation") continue;
		const std::string ps = g["parameterSet"].get<std::string>();
		for (const auto& t : g["tests"])
		{
			const std::string id = "tc" + std::to_string(t["tcId"].get<int>()) + " " + ps +
				" (" + t["reason"].get<std::string>() + ")";
			CK_OBJECT_HANDLE hPrk;
			CK_RV rv = importKey(hSession, true, mlkemParamSet(ps), acvpkat::hexField(t, "dk"), hPrk);
			if (rv != CKR_OK) { f.check(false, id + ": import dk rv=" + acvpkat::rvHex(rv)); continue; }
			CK_OBJECT_HANDLE hSecret;
			rv = decapsulate(hSession, hPrk, acvpkat::hexField(t, "c"), hSecret);
			if (rv != CKR_OK) { f.check(false, id + ": decapsulate rv=" + acvpkat::rvHex(rv)); }
			else
			{
				CK_BYTE k[64];
				CK_ATTRIBUTE a[] = { { CKA_VALUE, k, sizeof(k) } };
				rv = CRYPTOKI_F_PTR( C_GetAttributeValue(hSession, hSecret, a, 1) );
				const std::vector<unsigned char> expected = acvpkat::hexField(t, "k");
				f.check(rv == CKR_OK && a[0].ulValueLen == expected.size() &&
				        memcmp(k, expected.data(), expected.size()) == 0, id + ": shared secret differs from NIST k");
				CRYPTOKI_F_PTR( C_DestroyObject(hSession, hSecret) );
			}
			CRYPTOKI_F_PTR( C_DestroyObject(hSession, hPrk) );
		}
	}
	f.assertNone("mlkem_encapdecap_val_test.json decapsulation");
}

void MlKemInputCheckTests::testLengthBoundaries()
{
	CK_SESSION_HANDLE hSession;
	openSession(hSession);
	nlohmann::json doc = acvpkat::load("mlkem_encapdecap_val_test.json");
	const nlohmann::json* validDk = findCase(doc, "decapsulationKeyCheck", "ML-KEM-768", true);
	const nlohmann::json* validEk = findCase(doc, "encapsulationKeyCheck", "ML-KEM-768", true);
	const nlohmann::json* decap   = findCase(doc, "decapsulation", "ML-KEM-768", true);
	CPPUNIT_ASSERT(validDk != NULL && validEk != NULL && decap != NULL);
	const std::vector<unsigned char> dk = acvpkat::hexField(*validDk, "dk");
	const std::vector<unsigned char> ek = acvpkat::hexField(*validEk, "ek");
	CPPUNIT_ASSERT_EQUAL((size_t)2400, dk.size());
	CPPUNIT_ASSERT_EQUAL((size_t)1184, ek.size());

	acvpkat::Failures f;
	struct KeyCase { const char* what; bool isPrivate; std::vector<unsigned char> v; CK_ULONG ps; };
	std::vector<unsigned char> dkShort(dk.begin(), dk.end() - 1), dkLong(dk); dkLong.push_back(0);
	std::vector<unsigned char> ekShort(ek.begin(), ek.end() - 1), ekLong(ek); ekLong.push_back(0);
	const KeyCase keyCases[] = {
		{ "dk 1 byte short", true,  dkShort, CKP_ML_KEM_768 },
		{ "dk 1 byte long",  true,  dkLong,  CKP_ML_KEM_768 },
		{ "ek 1 byte short", false, ekShort, CKP_ML_KEM_768 },
		{ "ek 1 byte long",  false, ekLong,  CKP_ML_KEM_768 },
		{ "ML-KEM-768 dk labelled ML-KEM-1024", true,  dk, CKP_ML_KEM_1024 },
		{ "ML-KEM-768 ek labelled ML-KEM-512",  false, ek, CKP_ML_KEM_512 },
		{ "unknown parameter set",              false, ek, (CK_ULONG)0x7F },
	};
	for (const KeyCase& kc : keyCases)
	{
		CK_OBJECT_HANDLE h;
		CK_RV rv = importKey(hSession, kc.isPrivate, kc.ps, kc.v, h);
		f.check(rv == CKR_ATTRIBUTE_VALUE_INVALID && h == CK_INVALID_HANDLE,
		        std::string(kc.what) + ": expected CKR_ATTRIBUTE_VALUE_INVALID, got " + acvpkat::rvHex(rv));
		if (h != CK_INVALID_HANDLE) CRYPTOKI_F_PTR( C_DestroyObject(hSession, h) );
	}

	// Ciphertext lengths against an ML-KEM-768 key (valid c is 1088 bytes).
	CK_OBJECT_HANDLE hPrk;
	CK_RV rv = importKey(hSession, true, CKP_ML_KEM_768, acvpkat::hexField(*decap, "dk"), hPrk);
	CPPUNIT_ASSERT(rv == CKR_OK);
	const size_t ctLens[] = { 768 /* ML-KEM-512 */, 1568 /* ML-KEM-1024 */, 1087, 1089, 1 };
	for (size_t len : ctLens)
	{
		CK_OBJECT_HANDLE hSecret;
		rv = decapsulate(hSession, hPrk, std::vector<unsigned char>(len, 0x5a), hSecret);
		f.check(rv == CKR_WRAPPED_KEY_LEN_RANGE && hSecret == CK_INVALID_HANDLE,
		        std::to_string(len) + "-byte ciphertext vs ML-KEM-768 key: expected CKR_WRAPPED_KEY_LEN_RANGE, got " +
		        acvpkat::rvHex(rv));
		if (hSecret != CK_INVALID_HANDLE) CRYPTOKI_F_PTR( C_DestroyObject(hSession, hSecret) );
	}
	// ...and the right length still decapsulates.
	CK_OBJECT_HANDLE hSecret;
	rv = decapsulate(hSession, hPrk, acvpkat::hexField(*decap, "c"), hSecret);
	f.check(rv == CKR_OK, "1088-byte NIST ciphertext: rv=" + acvpkat::rvHex(rv));
	f.assertNone("ML-KEM length boundaries");
}

File: rust/src/mlkem_input_check_tests.rs
//! FIPS 203 §7.2 / §7.3 ML-KEM input checks at the PKCS#11 boundary
//! (2026-09-25).
//!
//! Defect: the engine accepted every NIST ACVP-Server encapDecap VAL
//! `encapsulationKeyCheck` / `decapsulationKeyCheck` case marked
//! `testPassed=false` (ek with a coefficient ≥ q; dk whose embedded
//! `h ≠ H(ek)`), and answered a wrong-length ciphertext with
//! CKR_ENCRYPTED_DATA_INVALID, a code PKCS#11 v3.2 §5.18.9 does not list.
//!
//! Vectors: `rust/kat/mlkem-keycheck-acvp.json` — byte-copied NIST
//! ACVP-Server@975de31e cases (provenance inside the file). Every assertion is
//! verdict/return-code/byte exact; nothing here is a round-trip-only check.
//!
//! Included from `ffi.rs` via `#[path]` so `use super::*` resolves to the FFI
//! module and its private helpers, like `conformance_v32_tests`.

use super::*;
use crate::native::test_lock;
use serde_json::Value;

/// High fixed session handle, disjoint from every other ffi test module.
const SESSION: u32 = 0x4D50_4001;

fn setup() {
    crate::state::set_initialized(true);
    SESSIONS.with(|s| {
        s.borrow_mut()
            .insert(SESSION, crate::state::SessionState { slot_id: 0, rw_session: true });
    });
    TOKEN_STORE.with(|ts| {
        ts.borrow_mut()
            .entry(0)
            .or_insert_with(|| crate::state::TokenState {
                slot_id: 0,
                initialized: true,
                label: [0u8; 32],
                login_state: crate::state::LoginState::User,
                so_pin_salt: [0u8; 16],
                so_pin_hash: [0u8; 32],
                user_pin_salt: None,
                user_pin_hash: None,
            })
            .login_state = crate::state::LoginState::User;
    });
}

fn kat() -> Value {
    serde_json::from_str(include_str!("../kat/mlkem-keycheck-acvp.json")).expect("parse kat")
}

fn unhex(s: &str) -> Vec<u8> {
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).expect("hex"))
        .collect()
}

fn ps_of(name: &str) -> u32 {
    match name {
        "ML-KEM-512" => CKP_ML_KEM_512,
        "ML-KEM-768" => CKP_ML_KEM_768,
        "ML-KEM-1024" => CKP_ML_KEM_1024,
        other => panic!("unknown parameter set {other}"),
    }
}

fn ct_len_of(ps: u32) -> usize {
    match ps {
        CKP_ML_KEM_512 => 768,
        CKP_ML_KEM_768 => 1088,
        CKP_ML_KEM_1024 => 1568,
        _ => unreachable!(),
    }
}

/// CK_ULONG-width little-endian bytes (what an LP64 caller sends).
fn ulong(v: u32) -> Vec<u8> {
    (v as usize).to_le_bytes().to_vec()
}

/// `C_CreateObject` through the real FFI entry point with a CK_ATTRIBUTE
/// array built from `entries`. Returns (rv, handle).
fn create(entries: Vec<(u32, Vec<u8>)>) -> (u32, u32) {
    let mut words: Vec<usize> = Vec::with_capacity(entries.len() * 3);
    for (t, v) in &entries {
        words.push(*t as usize);
        words.push(v.as_ptr() as usize);
        words.push(v.len());
    }
    let mut h: u32 = 0;
    let rv = C_CreateObject(SESSION, words.as_mut_ptr() as *mut u8, entries.len() as u32, &mut h);
    (rv, h)
}

fn import_ek(ps: u32, ek: &[u8]) -> (u32, u32) {
    create(vec![
        (CKA_CLASS, ulong(CKO_PUBLIC_KEY)),
        (CKA_KEY_TYPE, ulong(CKK_ML_KEM)),
        (CKA_PARAMETER_SET, ulong(ps)),
        (CKA_ENCAPSULATE, vec![1]),
        (CKA_TOKEN, vec![0]),
        (CKA_VALUE, ek.to_vec()),
    ])
}

fn import_dk(ps: u32, dk: &[u8]) -> (u32, u32) {
    create(vec![
        (CKA_CLASS, ulong(CKO_PRIVATE_KEY)),
        (CKA_KEY_TYPE, ulong(CKK_ML_KEM)),
        (CKA_PARAMETER_SET, ulong(ps)),
        (CKA_DECAPSULATE, vec![1]),
        (CKA_TOKEN, vec![0]),
        (CKA_PRIVATE, vec![0]),
        (CKA_VALUE, dk.to_vec()),
    ])
}

fn encapsulate(h_pub: u32, ct_buf_len: usize) -> (u32, u32) {
    let mut mech: [usize; 3] = [CKM_ML_KEM as usize, 0, 0];
    let mut ct = vec![0u8; ct_buf_len];
    let mut ct_len = ct_buf_len as u32;
    let mut h_new: u32 = 0;
    let rv = C_EncapsulateKey(
        SESSION,
        mech.as_mut_ptr() as *mut u8,
        h_pub,
        std::ptr::null_mut(),
        0,
        ct.as_mut_ptr(),
        &mut ct_len,
        &mut h_new,
    );
    (rv, h_new)
}

fn decapsulate(h_prv: u32, ct: &[u8]) -> (u32, u32) {
    let mut mech: [usize; 3] = [CKM_ML_KEM as usize, 0, 0];
    let mut ct = ct.to_vec();
    let mut h_new: u32 = 0;
    let rv = C_DecapsulateKey(
        SESSION,
        mech.as_mut_ptr() as *mut u8,
        h_prv,
        std::ptr::null_mut(),
        0,
        ct.as_mut_ptr(),
        ct.len() as u32,
        &mut h_new,
    );
    (rv, h_new)
}

fn destroy(h: u32) {
    if h != 0 {
        OBJECTS.with(|o| o.borrow_mut().remove(&h));
    }
}

/// The helper verdicts themselves, against NIST's `testPassed` — all 12
/// key-check cases (6 valid, 6 invalid), no FFI in between.
#[test]
fn fips203_key_check_helpers_match_nist_verdicts() {
    let doc = kat();
    let cases = doc["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 12, "fixture must carry all 12 key-check cases");
    let mut invalid_seen = 0;
    for c in cases {
        let ps = ps_of(c["parameterSet"].as_str().unwrap());
        let want = c["testPassed"].as_bool().unwrap();
        let got = match c["function"].as_str().unwrap() {
            "encapsulationKeyCheck" => {
                crate::native::keygen::ml_kem_ek_check(ps, &unhex(c["ek"].as_str().unwrap()))
            }
            "decapsulationKeyCheck" => {
                crate::native::keygen::ml_kem_dk_check(ps, &unhex(c["dk"].as_str().unwrap()))
            }
            other => panic!("unexpected function {other}"),
        };
        assert_eq!(got, Some(want), "tc{} ({}) verdict", c["tcId"], c["reason"]);
        if !want {
            invalid_seen += 1;
        }
    }
    assert_eq!(invalid_seen, 6);
}

/// FIPS 203 §7.2/§7.3 through `C_CreateObject` (PKCS#11 v3.2 §4.1.1 rule 2):
/// a NIST-invalid key → exactly CKR_ATTRIBUTE_VALUE_INVALID, no object; a
/// NIST-valid key → CKR_OK, and the key then works under C_EncapsulateKey /
/// C_DecapsulateKey (all-zero ciphertext of the right length: ML-KEM's
/// implicit rejection returns CKR_OK with a 32-byte secret).
#[test]
fn nist_key_checks_through_create_object() {
    let _guard = test_lock::acquire();
    setup();
    let doc = kat();
    for c in doc["cases"].as_array().unwrap() {
        let ps = ps_of(c["parameterSet"].as_str().unwrap());
        let valid = c["testPassed"].as_bool().unwrap();
        let tag = format!("tc{} {} ({})", c["tcId"], c["parameterSet"], c["reason"]);
        let is_ek = c["function"] == "encapsulationKeyCheck";
        let (rv, h) = if is_ek {
            import_ek(ps, &unhex(c["ek"].as_str().unwrap()))
        } else {
            import_dk(ps, &unhex(c["dk"].as_str().unwrap()))
        };
        if !valid {
            assert_eq!(rv, CKR_ATTRIBUTE_VALUE_INVALID, "{tag}: must be rejected at import");
            assert_eq!(h, 0, "{tag}: no object on failure");
            continue;
        }
        assert_eq!(rv, CKR_OK, "{tag}: NIST-valid key must import");
        let (rv2, h_ss) = if is_ek {
            encapsulate(h, ct_len_of(ps))
        } else {
            decapsulate(h, &vec![0u8; ct_len_of(ps)])
        };
        assert_eq!(rv2, CKR_OK, "{tag}: NIST-valid key must be usable");
        assert_eq!(get_object_value(h_ss).map(|v| v.len()), Some(32), "{tag}");
        destroy(h_ss);
        destroy(h);
    }
}

/// A NIST-valid dk still decapsulates BYTE-EXACTLY after the checks: the
/// shared secret equals NIST's `k` for the lowest-tcId "valid decapsulation"
/// case of each parameter set.
#[test]
fn nist_valid_decapsulation_is_byte_exact() {
    let _guard = test_lock::acquire();
    setup();
    let doc = kat();
    let cases = doc["decapsulation_cases"].as_array().unwrap();
    assert_eq!(cases.len(), 3);
    for c in cases {
        let ps = ps_of(c["parameterSet"].as_str().unwrap());
        let (rv, h) = import_dk(ps, &unhex(c["dk"].as_str().unwrap()));
        assert_eq!(rv, CKR_OK, "tc{}", c["tcId"]);
        let (rv2, h_ss) = decapsulate(h, &unhex(c["c"].as_str().unwrap()));
        assert_eq!(rv2, CKR_OK, "tc{}", c["tcId"]);
        assert_eq!(
            get_object_value(h_ss),
            Some(unhex(c["k"].as_str().unwrap())),
            "tc{}: shared secret must equal NIST k",
            c["tcId"]
        );
        destroy(h_ss);
        destroy(h);
    }
}

/// Use-time backstop: a NIST-invalid key that reached the store WITHOUT
/// C_CreateObject (as C_UnwrapKey or a pre-fix persisted object would) must
/// still never be run through ML-KEM.Encaps / ML-KEM.Decaps.
#[test]
fn invalid_keys_injected_into_store_are_refused_at_use() {
    let _guard = test_lock::acquire();
    setup();
    let doc = kat();
    let mut handle = 0x4D50_4100u32;
    for c in doc["cases"].as_array().unwrap() {
        if c["testPassed"].as_bool().unwrap() {
            continue;
        }
        let ps = ps_of(c["parameterSet"].as_str().unwrap());
        let is_ek = c["function"] == "encapsulationKeyCheck";
        handle += 1;
        OBJECTS.with(|o| {
            let mut attrs = Attributes::new();
            if is_ek {
                attrs.insert(CKA_VALUE, unhex(c["ek"].as_str().unwrap()));
                store_ulong(&mut attrs, CKA_CLASS, CKO_PUBLIC_KEY);
                store_bool(&mut attrs, CKA_ENCAPSULATE, true);
            } else {
                attrs.insert(CKA_VALUE, unhex(c["dk"].as_str().unwrap()));
                store_ulong(&mut attrs, CKA_CLASS, CKO_PRIVATE_KEY);
                store_bool(&mut attrs, CKA_DECAPSULATE, true);
            }
            store_ulong(&mut attrs, CKA_KEY_TYPE, CKK_ML_KEM);
            store_param_set(&mut attrs, ps);
            o.borrow_mut().insert(handle, attrs);
        });
        let (rv, h_ss) = if is_ek {
            encapsulate(handle, ct_len_of(ps))
        } else {
            decapsulate(handle, &vec![0u8; ct_len_of(ps)])
        };
        assert_eq!(rv, CKR_KEY_TYPE_INCONSISTENT, "tc{} ({})", c["tcId"], c["reason"]);
        assert_eq!(h_ss, 0, "tc{}: no key object on failure", c["tcId"]);
        destroy(handle);
    }
}

/// Length boundaries, ML-KEM-512 (FIPS 203 §7.2 / §7.3 type checks):
/// ek/dk one byte short or long → CKR_ATTRIBUTE_VALUE_INVALID at
/// C_CreateObject; ciphertext one byte short/long or another parameter set's
/// length → CKR_WRAPPED_KEY_LEN_RANGE (PKCS#11 v3.2 §5.18.9 list; §5.1.6
/// "invalid solely on the basis of its length"), never a key object.
#[test]
fn length_boundaries_have_spec_return_codes() {
    let _guard = test_lock::acquire();
    setup();
    let doc = kat();
    let cases = doc["cases"].as_array().unwrap();
    let pick = |f: &str| {
        cases
            .iter()
            .find(|c| c["parameterSet"] == "ML-KEM-512" && c["function"] == f && c["testPassed"] == true)
            .unwrap()
    };
    let ek = unhex(pick("encapsulationKeyCheck")["ek"].as_str().unwrap());
    let dk = unhex(pick("decapsulationKeyCheck")["dk"].as_str().unwrap());
    assert_eq!((ek.len(), dk.len()), (800, 1632));
    let ps = CKP_ML_KEM_512;

    let mut ek_long = ek.clone();
    ek_long.push(0);
    for (label, bad) in [("ek-1", ek[..ek.len() - 1].to_vec()), ("ek+1", ek_long)] {
        assert_eq!(import_ek(ps, &bad), (CKR_ATTRIBUTE_VALUE_INVALID, 0), "{label}");
    }
    let mut dk_long = dk.clone();
    dk_long.push(0);
    for (label, bad) in [("dk-1", dk[..dk.len() - 1].to_vec()), ("dk+1", dk_long)] {
        assert_eq!(import_dk(ps, &bad), (CKR_ATTRIBUTE_VALUE_INVALID, 0), "{label}");
    }

    let (rv, h_prv) = import_dk(ps, &dk);
    assert_eq!(rv, CKR_OK);
    for (label, len) in [("ct-1", 767usize), ("ct+1", 769), ("ct=768-set", 1088), ("ct=0", 0)] {
        let buf = vec![0x5Au8; len.max(1)];
        let (rv, h_ss) = decapsulate(h_prv, &buf[..len]);
        assert_eq!(rv, CKR_WRAPPED_KEY_LEN_RANGE, "{label}");
        assert_eq!(h_ss, 0, "{label}: no key object on failure");
    }
    let (rv, h_ss) = decapsulate(h_prv, &vec![0u8; 768]);
    assert_eq!(rv, CKR_OK, "exact length must be accepted");
    destroy(h_ss);
    destroy(h_prv);
}

/// The KMIP ingress (`native::register_ml_kem_{private,public}_key`) applies
/// the same checks: NIST-invalid → CKR_ATTRIBUTE_VALUE_INVALID, valid → Ok.
#[test]
fn native_register_applies_the_same_checks() {
    let _guard = test_lock::acquire();
    let _ = crate::native::session::finalize();
    crate::native::session::init().expect("engine init");
    let session = crate::native::session::bootstrap_default_token(0, "so", "user", "mlkem-check")
        .expect("bootstrap session");
    let doc = kat();
    for c in doc["cases"].as_array().unwrap() {
        let ps = ps_of(c["parameterSet"].as_str().unwrap());
        let valid = c["testPassed"].as_bool().unwrap();
        let r = if c["function"] == "encapsulationKeyCheck" {
            crate::native::keygen::register_ml_kem_public_key(
                session,
                ps,
                &unhex(c["ek"].as_str().unwrap()),
                b"ek",
                "check",
            )
        } else {
            crate::native::keygen::register_ml_kem_private_key(
                session,
                ps,
                &unhex(c["dk"].as_str().unwrap()),
                b"dk",
                "check",
            )
        };
        if valid {
            assert!(r.is_ok(), "tc{}: valid key must register", c["tcId"]);
        } else {
            assert_eq!(r, Err(CKR_ATTRIBUTE_VALUE_INVALID), "tc{} ({})", c["tcId"], c["reason"]);
        }
    }
    let _ = crate::native::session::finalize();
}
